### DISCUSSION TITLE
iperf3.16 UDP test acceptance issue

### DISCUSSION THREAD

**Discussion by hzywhite** [esnet/iperf#1674] [Q&A]
I tested version 3.16 of iperfs on two Windows 11 computers and during the UDP packet filling process
why the receiving end's rate kept showing 0

server:iperf3 -u -c 192.168.43.100 -i 1 -l 1470 -b 200m -t 999
client:iperf3 -s -i 1

**Comment by davidBar-On** [comment]
@hzywhite, can you give more details?  (Note by the way that you tagged the server command line as "client" and visa versa.)

1.  Does the server shows receiving rate 0 from the first interval (fist second) or it happens only sometime during the test?
2. When the server receive rate is 0, does the client continue to send the 200Mbits/sec?
3. Can you run both server and client with the additional options `-V --debug=3` and include the first part of this output and a part from when the issue happens?

**Reply by hzywhite** [reply]
warning: UDP block size 1470 exceeds TCP MSS 1460, may result in fragmentation / drops
Connecting to host 192.168.43.102, port 5201
[  5] local 192.168.43.10 port 50337 connected to 192.168.43.102 port 5201
[ ID] Interval           Transfer     Bitrate         Total Datagrams
[  5]   0.00-1.00   sec  23.6 MBytes   197 Mbits/sec  16815
[  5]   1.00-2.01   sec  24.1 MBytes   200 Mbits/sec  17195
[  5]   2.01-3.01   sec  23.9 MBytes   200 Mbits/sec  17048
[  5]   3.01-4.01   sec  23.8 MBytes   200 Mbits/sec  16950
[  5]   4.01-5.00   sec  23.7 MBytes   200 Mbits/sec  16899
[  5]   5.00-6.02   sec  24.1 MBytes   200 Mbits/sec  17180
[  5]   6.02-7.01   sec  23.7 MBytes   200 Mbits/sec  16931
[  5]   7.01-8.01   sec  23.8 MBytes   200 Mbits/sec  16977
[  5]   8.01-9.01   sec  23.8 MBytes   200 Mbits/sec  16945
[  5]   9.01-10.01  sec  24.0 MBytes   200 Mbits/sec  17143
[  5]  10.01-11.00  sec  23.5 MBytes   200 Mbits/sec  16796
[  5]  11.00-12.01  sec  24.1 MBytes   200 Mbits/sec  17217
[  5]  12.01-13.01  sec  23.7 MBytes   200 Mbits/sec  16899
[  5]  13.01-14.00  sec  23.8 MBytes   200 Mbits/sec  16961


[ ID] Interval           Transfer     Bitrate         Jitter    Lost/Total Datagrams
[  5]   0.00-1.00   sec  19.1 MBytes   160 Mbits/sec  0.115 ms  3163/16815 (19%)
[  5]   1.00-2.00   sec   488 KBytes  4.01 Mbits/sec  0.017 ms  20/360 (5.6%)
[  5]   2.00-3.00   sec  0.00 Bytes  0.00 bits/sec  0.017 ms  0/0 (0%)
[  5]   3.00-4.00   sec  0.00 Bytes  0.00 bits/sec  0.017 ms  0/0 (0%)
[  5]   4.00-5.00   sec  0.00 Bytes  0.00 bits/sec  0.017 ms  0/0 (0%)
[  5]   5.00-6.00   sec  0.00 Bytes  0.00 bits/sec  0.017 ms  0/0 (0%)
[  5]   6.00-7.00   sec  0.00 Bytes  0.00 bits/sec  0.017 ms  0/0 (0%)
[  5]   7.00-8.00   sec  0.00 Bytes  0.00 bits/sec  0.017 ms  0/0 (0%)
[  5]   8.00-9.00   sec  0.00 Bytes  0.00 bits/sec  0.017 ms  0/0 (0%)
[  5]   9.00-10.00  sec  0.00 Bytes  0.00 bits/sec  0.017 ms  0/0 (0%)
[  5]  10.00-11.00  sec  0.00 Bytes  0.00 bits/sec  0.017 ms  0/0 (0%)
[  5]  11.00-12.00  sec  2.65 MBytes  22.2 Mbits/sec  0.029 ms  170483/172370 (99%)
[  5]  12.00-13.00  sec  0.00 Bytes  0.00 bits/sec  0.029 ms  0/0 (0%)
[  5]  13.00-14.00  sec  0.00 Bytes  0.00 bits/sec  0.029 ms  0/0 (0%)

**Reply by hzywhite** [reply]
Regarding the third point, I tested it and found that adding - V to the client made reception appear normal with slight packet loss. At first, I thought it was caused by a system interrupt, so I bound the CPU, but the effect was that it couldn't handle it, resulting in a large amount of packet loss.
It seems that this slight packet loss is related to the priority of the process...

iperf3 -s -i 1 -V
iperf 3.16
CYGWIN_NT-10.0-19043 HG7111007875 3.4.10-1.x86_64 2023-11-29 12:12 UTC x86_64
-----------------------------------------------------------
Server listening on 5201 (test #1)
-----------------------------------------------------------
Time: Tue, 26 Mar 2024 11:15:35 GMT
Accepted connection from 192.168.43.10, port 50311
      Cookie: 67s5vqoc3lzid5oxl3i3a3chaoicwo73zkpc
      Target Bitrate: 200000000
[  5] local 192.168.43.102 port 5201 connected to 192.168.43.10 port 50372
Starting Test: protocol: UDP, 1 streams, 1470 byte blocks, omitting 0 seconds, 999 second test, tos 0
[ ID] Interval           Transfer     Bitrate         Jitter    Lost/Total Datagrams
[  5]   0.00-1.00   sec  23.4 MBytes   196 Mbits/sec  0.011 ms  188/16864 (1.1%)
[  5]   1.00-2.01   sec  23.5 MBytes   196 Mbits/sec  0.009 ms  258/17053 (1.5%)
[  5]   2.01-3.00   sec  23.6 MBytes   199 Mbits/sec  0.063 ms  90/16918 (0.53%)
[  5]   3.00-4.00   sec  23.5 MBytes   197 Mbits/sec  0.082 ms  233/16968 (1.4%)
[  5]   4.00-5.01   sec  23.7 MBytes   198 Mbits/sec  0.087 ms  202/17134 (1.2%)
[  5]   5.01-6.00   sec  23.6 MBytes   200 Mbits/sec  0.014 ms  144/16988 (0.85%)
[  5]   6.00-7.00   sec  23.6 MBytes   198 Mbits/sec  0.073 ms  108/16931 (0.64%)
[  5]   7.00-8.00   sec  23.5 MBytes   198 Mbits/sec  0.074 ms  179/16967 (1.1%)


bound the CPU：
[ ID] Interval           Transfer     Bitrate         Jitter    Lost/Total Datagrams
[  5]   0.00-1.00   sec  23.1 MBytes   193 Mbits/sec  0.071 ms  348/16791 (2.1%)
[  5]   1.00-2.00   sec  23.6 MBytes   198 Mbits/sec  0.009 ms  205/17033 (1.2%)
[  5]   2.00-3.00   sec  23.7 MBytes   199 Mbits/sec  0.063 ms  104/17006 (0.61%)
[  5]   3.00-4.00   sec  23.7 MBytes   199 Mbits/sec  0.070 ms  105/17021 (0.62%)
[  5]   4.00-5.00   sec  23.8 MBytes   200 Mbits/sec  0.009 ms  100/17068 (0.59%)
[  5]   5.00-6.00   sec  23.8 MBytes   199 Mbits/sec  0.009 ms  110/17057 (0.64%)
[  5]   6.00-7.01   sec  23.5 MBytes   196 Mbits/sec  0.075 ms  176/16973 (1%)
[  5]   7.01-8.00   sec  23.5 MBytes   199 Mbits/sec  0.066 ms  120/16907 (0.71%)
[  5]   8.00-9.00   sec  23.7 MBytes   199 Mbits/sec  0.073 ms  92/16965 (0.54%)
[  5]   9.00-10.00  sec  4.55 MBytes  38.2 Mbits/sec  0.152 ms  12516/15760 (79%)
[  5]  10.00-11.00  sec  2.42 MBytes  20.3 Mbits/sec  0.340 ms  15966/17693 (90%)
[  5]  11.00-12.00  sec  2.86 MBytes  24.0 Mbits/sec  1.503 ms  15358/17401 (88%)
[  5]  12.00-13.00  sec  1.73 MBytes  14.5 Mbits/sec  0.206 ms  15483/16716 (93%)
[  5]  13.00-14.00  sec  4.11 MBytes  34.5 Mbits/sec  0.018 ms  14560/17493 (83%)
[  5]  14.00-15.01  sec  19.1 MBytes   159 Mbits/sec  0.087 ms  3459/17097 (20%)
[  5]  15.01-16.00  sec  19.3 MBytes   162 Mbits/sec  0.013 ms  3294/17042 (19%)
[  5]  16.00-17.01  sec  18.8 MBytes   158 Mbits/sec  0.086 ms  3520/16950 (21%)
[  5]  17.01-18.00  sec  6.86 MBytes  57.9 Mbits/sec  0.016 ms  1091/5987 (18%)
[  5]  18.00-19.00  sec  0.00 Bytes  0.00 bits/sec  0.016 ms  0/0 (0%)
[  5]  19.00-20.00  sec  0.00 Bytes  0.00 bits/sec  0.016 ms  0/0 (0%)
[  5]  20.00-21.00  sec  0.00 Bytes  0.00 bits/sec  0.016 ms  0/0 (0%)
[  5]  20.00-21.00  sec  0.00 Bytes  0.00 bits/sec  0.016 ms  0/0 (0%)

**Comment by davidBar-On** [comment]
This is strange.  I don't know of any reason that `-V` will make a difference.  Try adding `-R` to see if the issue also happens when the server is sending to the client.

Try running a TCP test without bitrate limit to see what is the network throughput and how it behaves.

Other UDP tests that may help understanding the issue changing to `-l 1460` (or even `-l 512`), so no fragmentation will be done, and try using different bit-rates. 

Running Wireshark at the network connection points to the client and the server may help to determine whether the client machine is loosing the packets, whether they are lost in the way, or whether the handling in the server is the problem.
